src/Protocols/DfLike.hs

```diff
@@ -41,931 +68,1546 @@ module Protocols.DfLike
+import           Protocols.Axi4.WriteData
+import           Protocols.Axi4.WriteResponse
+import           Protocols.Df (Data(..), Df)
+import qualified Protocols.Df as Df
+import           Protocols.Internal
+
+
+-- | Class for protocols that are "similar" to 'Df',
+-- i.e. they can be converted into a 'Df' port using a 'Circuit' (see 'toDfCircuit').
+-- This is for protocols that carry some "interesting" data,
+-- as well as some "uninteresting" data (e.g. address, burst length).
+-- The 'Circuit' should abstract away the complexities of each protocol,
+-- so that they can be dealt with uniformly using 'Df'.
+-- For pipelined protocols, which can carry both in the same cycle,
+-- the 'Circuit' should pass along the interesting parts
+-- but not the uninteresting parts.
+--
+-- Can take parameters, e.g. for addresses.
+-- Defaults to being the right side of the circuit (@Fwd ~ Otp, Bwd ~ Inp@),
+-- but this can be switched using 'Reverse' from 'Protocols.Internal'.
+-- Supports both bwd (input/read) data and fwd (output/write) data.
+class (Protocol df) => DfLike df where
+  -- | Domain that messages are being sent over.
+  -- It should be true that @Fwd df ~ Signal dom [something]@
+  -- and that @Bwd df ~ Signal dom [something]@.
+  type Dom df :: Domain
+  -- | Information being sent into the port, along @Bwd df@.
+  -- This is the information being carried over the protocol,
+  -- /not/ the messages being carried over the protocol,
+  -- so it doesn't include auxiliary information like address or burst length.
+  -- If no data is sent in this direction, set this to @()@.
+  type BwdPayload df
+  -- | Information being sent out from the port, along @Fwd df@.
+  -- This is the information being carried over the protocol,
+  -- /not/ the messages being carried over the protocol,
+  -- so it doesn't include auxiliary information like address or burst length.
+  -- If no data is sent in this direction, set this to @()@.
+  type FwdPayload df
+  -- | User-provided parameters for 'toDfCircuit'
+  -- (e.g. address to respond to, so that different 'DfLike'
+  -- components can respond to different addresses).
+  -- If you don't need to take params, set this to @()@.
+  type DfLikeParam df
+  -- | Circuit which converts Df into this protocol's messages.
+  -- This should deal with all the complexities of your protocol
+  -- such as addresses, bursts, pipelining, etc. so that
+  -- a circuit connected to the 'Df' end doesn't have to worry about all that.
+  -- There are two Df channels, one for fwd data and one for bwd data,
+  -- so data can be sent both ways at once.
+  -- This circuit is expected to follow all of the conventions of 'Df';
+  -- for example, 'Df.Data' should stay the same
+  -- between clock cycles unless acknowledged.
+  toDfCircuit :: (HiddenClockResetEnable (Dom df))
+    => (Proxy df, DfLikeParam df)
+    -> Circuit (Df (Dom df) (FwdPayload df), Reverse (Df (Dom df) (BwdPayload df))) df
```

Comment:
```suggestion
  toDfCircuit :: 
    (HiddenClockResetEnable (Dom df)) =>
    (Proxy df, DfLikeParam df) ->
    Circuit (Df (Dom df) (FwdPayload df), Reverse (Df (Dom df) (BwdPayload df))) df
```

main motivation for this style is that we'll get even more type symbols if we ever start using linear types. It also makes copy+pasting a bit better, as [explained here](https://yairchu.github.io/posts/silly-haskell-formatting).

Reply:
This would need to be applied in other places in this PR too.